internal/prepare/prepare.go

```diff
@@ -0,0 +1,76 @@
+package prepare
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/canonical/sqlair/internal/parse"
+	"github.com/canonical/sqlair/internal/typeinfo"
+)
+
+// PreparedExpr represents an SQL expression after the input and output parts
+// have been replaced by their corresponding expansions.
+type PreparedExpr struct {
+	ParsedExpr *parse.ParsedExpr
+	SQL        string
+}
+
+type typeNameToInfo map[string]*typeinfo.Info
+
+// prepareInput checks that the input expression corrosponds to a Go struct
+// passed to Prepare.
+func prepareInput(ti typeNameToInfo, p *parse.InputPart) error {
+	if inf, ok := ti[p.Source.Prefix]; ok {
+		if _, ok := inf.TagToField[p.Source.Name]; ok {
+			return nil
+		}
+		return fmt.Errorf("there is no tag with name %s in %s",
+			p.Source.Name, inf.Type.Name())
+	}
+	return fmt.Errorf("unknown type: %s", p.Source.Prefix)
+}
+
+// Prepare takes a parsed expression and all the go objects mentioned in it.
+// The IO parts of the statement are checked for validity against the Go objects
+// and expanded if necessary.
+func Prepare(pe *parse.ParsedExpr, args ...any) (expr *PreparedExpr, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot prepare expression: %s", err)
+		}
+	}()
+
+	var ti = make(typeNameToInfo)
+
+	// Generate and save reflection info.
+	for _, arg := range args {
+		i, err := typeinfo.TypeInfo(arg)
+		if err != nil {
+			return nil, err
+		}
+		ti[i.Type.Name()] = i
+	}
+
+	var sql bytes.Buffer
+	// Check and expand each query part.
+	for _, part := range pe.QueryParts {
+		if p, ok := part.(*parse.InputPart); ok {
+			err := prepareInput(ti, p)
+			if err != nil {
+				return nil, err
+			}
+			sql.WriteString(p.ToSQL())
+			continue
+		}
+		if p, ok := part.(*parse.OutputPart); ok {
+			// Do nothing for now.
+			sql.WriteString(p.ToSQL())
+			continue
+		}
+		p := part.(*parse.BypassPart)
+		sql.WriteString(p.ToSQL())
+	}
+
+	return &PreparedExpr{ParsedExpr: pe, SQL: strings.TrimSpace(sql.String())}, nil
```

Comment:
True, this can actually be removed. Previously we had been enforcing correct spacing around queries, but now its been changed to accept whatever was there before. The `TrimSpace` took care of spaces round IO expressions at the start of end of the string.